_posts/2020-06-18-life-locks.md

```diff
@@ -0,0 +1,155 @@
+
+### Third Requirement
+
+As the exact legislative requirements around retention haven't been decided, the implementation needs to be flexible. You know you will need to react quickly when the rules change, and [Bucket Policies](https://docs.aws.amazon.com/AmazonS3/latest/dev/access-policy-language-overview.html) paired with [Conditions Keys](https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_policies_elements_condition.html) grant you this freedom. You can continually make code changes and redeploy as often as required by using these two condition keys:
+
+1. [s3:object-lock-mode](https://docs.aws.amazon.com/AmazonS3/latest/dev/list_amazons3.html#amazons3-policy-keys) is used to enforce which locking mode is used; and
+2. [s3:object-lock-remaining-retention-days](https://docs.aws.amazon.com/AmazonS3/latest/dev/list_amazons3.html#amazons3-policy-keys) is used to enforce the required lock duration.
+
+#### s3:object-lock-mode
+
+By using `s3:object-lock-mode` you can enforce which [S3 Actions](https://docs.aws.amazon.com/IAM/latest/UserGuide/list_amazons3.html#amazons3-actions-as-permissions) require a lock mode value (the first stanza), as well as the requirement that it has a value of **COMPLIANCE** (second stanza).
+
+```yaml
+- Sid: 'RequireObjectLockMode'
+  Effect: Deny
+  Principal: '*'
+  Action:
+    - 's3:PutObject'
+    - 's3:PutObjectRetention'
+  Resource:
+    - !Sub '${Bucket.Arn}/*'
+  Condition:
+      'Null':
+        's3:object-lock-mode': 'true'
+- Sid: 'RequireObjectLockModeToBeCompliance'
+  Effect: Deny
+  Principal: '*'
+  Action:
+    - 's3:PutObject'
+    - 's3:PutObjectRetention'
+  Resource:
+    - !Sub '${Bucket.Arn}/*'
+  Condition:
+      'StringNotEqualsIgnoreCase':
+        's3:object-lock-mode': 'COMPLIANCE'
+```
+
+#### s3:object-lock-remaining-retention-days
+
+Next you enforce how long uploaded files are protected by using the `s3:object-lock-remaining-retention-days` condition key. In the example below, you require that a retention period be specified, and that it must be 7 days. It is also possible to use this condition key to enforce a day range; I'll leave that as an exercise for you dear reader :wink:
+
+```yaml
+- Sid: 'RequireRetentionDays'
+  Effect: Deny
+  Principal: '*'
+  Action:
+    - 's3:PutObject'
+    - 's3:PutObjectRetention'
+  Resource:
+    - !Sub '${Bucket.Arn}/*'
+  Condition:
+      'Null':
+        's3:object-lock-remaining-retention-days': 'true'
+- Sid: 'RequireRetentionToBeThreeDays'
+  Effect: Deny
+  Principal: '*'
+  Action:
+    - 's3:PutObject'
+    - 's3:PutObjectRetention'
+  Resource:
+    - !Sub '${Bucket.Arn}/*'
+  Condition:
+      NumericNotEquals:
+        's3:object-lock-remaining-retention-days': '7'
+```
+
+### Unspoken Requirement
+
+The last and unspoken requirement can be found in this [complete CloudFormation template](/assets/2020-06-03-life-locks-example-template.yml).
```

Comment:
This will open a tab and most likely download the file. Is that the intended behaviour? If you want the user to be able to browse this you can always create a github "gist", for example: https://gist.github.com/jgunnink/00714353f09855220cc54c0132f88cd9

Just thinking for mobile users or people who want to just quickly view without downloading. Up to you 😁

Reply:
Good idea; I've actually added a Gist link and kept in the file :)